File: src/app-context.ts
import os from 'node:os';
import path from 'node:path';
import { systemClock } from './infrastructure/clock.js';
import { createDatabaseProvider } from './infrastructure/database.js';
import { createCliStateStore } from './infrastructure/state-store.js';
import { createMcpListingRepository } from './repositories/mcp-listing-repository.js';
import { createSessionRepository } from './repositories/session-repository.js';
import { createUserRepository } from './repositories/user-repository.js';
import { createAuthService } from './services/auth-service.js';
import { createMcpService } from './services/mcp-service.js';

export interface AppContext {
  authService: ReturnType<typeof createAuthService>;
  mcpService: ReturnType<typeof createMcpService>;
  cliStateStore: ReturnType<typeof createCliStateStore>;
}

export function resolveAppDataDirectory(baseDir?: string): string {
  if (baseDir) {
    return path.join(baseDir, '.local');
  }

  const xdgDataHome = process.env.XDG_DATA_HOME?.trim();
  if (xdgDataHome) {
    return path.join(xdgDataHome, 'mcp-marketplace');
  }

  return path.join(os.homedir(), '.local', 'share', 'mcp-marketplace');
}

export function createAppContext(baseDir?: string): AppContext {
  const dataDir = resolveAppDataDirectory(baseDir);
  const database = createDatabaseProvider(path.join(dataDir, 'marketplace.db'));
  database.runMigrations();

  const users = createUserRepository(database.db);
  const sessions = createSessionRepository(database.db);
  const listings = createMcpListingRepository(database.db);

  return {
    authService: createAuthService({ users, sessions, clock: systemClock }),
    mcpService: createMcpService({ listings, clock: systemClock }),
    cliStateStore: createCliStateStore(path.join(dataDir, 'cli-state.json')),
  };
}

File: src/app_context.py
import os
from pathlib import Path

from src.infrastructure.database import create_connection, run_migrations
from src.infrastructure.state_store import CliStateStore
from src.repositories.mcp_listing_repository import McpListingRepository
from src.repositories.session_repository import SessionRepository
from src.repositories.user_repository import UserRepository
from src.services.auth_service import AuthService
from src.services.mcp_service import McpService


def resolve_app_data_directory(base_dir: str | None = None) -> Path:
    if base_dir:
        return Path(base_dir) / "OpenKG-McpMarket"

    xdg_data_home = os.environ.get("XDG_DATA_HOME", "").strip()
    if xdg_data_home:
        return Path(xdg_data_home) / "OpenKG-McpMarket"

    return Path.home() / ".local" / "share" / "OpenKG-McpMarket"


class AppContext:
    def __init__(self, base_dir: str | None = None) -> None:
        data_dir = resolve_app_data_directory(base_dir)
        connection = create_connection(data_dir / "marketplace.db")
        run_migrations(connection)

        users = UserRepository(connection)
        sessions = SessionRepository(connection)
        listings = McpListingRepository(connection)

        self.auth_service = AuthService(users, sessions)
        self.mcp_service = McpService(listings)
        self.cli_state_store = CliStateStore(data_dir / "cli-state.json")
